Made some changes to reference parsing

sortofbusy opened the pull request:
First of all, I appreciate your work.

My changes:
- handle abbreviation periods in references
- enable the input string to be any case
- fix an indexing bug on the toString() function ["Exodus 1:1" was returning Numbers 1:1]
- enable toString() to display a reference with no verse component

Also, the npm page for this project still lists the package as "bible" instead of "biblejs".

davewasmer commented on lib/reference.coffee:
```diff
@@ -37,7 +38,9 @@ class Reference
     # (e.g. John 1) then @verse is undefined
-    [@chapter, @verse] = chapterAndVerse.split(':')
+    parts = chapterAndVerse.split(':')
+    @chapter = +parts[0]
```
Could you use something like:

```coffee
[ @chapter, @verse ] = chapterAndVerse.split(':')
@chapter = Number(@chapter)
@verse = Number(@verse) or null
```

I'd like to avoid the `variable = +variable` pattern for type coercion.

Also, any reason why you used `-1` to indicate no verse? I'd prefer to keep it more explicit - either null or undefined.

sortofbusy replied:
Made this change. I just used -1 out of old habit.

davewasmer commented on lib/reference.coffee:
```diff
@@ -59,20 +62,19 @@ class Reference
-    verseNumber = @verse
-    return "#{bookName} #{chapterNumber}:#{verseNumber}"
+    verseNumber = if @verse > -1 then ":" + @verse else ""
+    return "#{bookName} #{chapterNumber}#{verseNumber}"
```
Could you make this a bit clearer by splitting it out? Perhaps:

```coffee
stringified = bookName + " " + chapterNumber
stringified += ":" + @verse if @verse
return stringified
```

sortofbusy replied:
Made this change, and took out the chapterNumber assignment line

davewasmer commented on lib/reference.coffee:
```diff
@@ -59,20 +62,19 @@ class Reference
-      count += books[@book].verses[i - 1]
-    count += @verse if @verse?
+      count += books[@book - 1].verses[i - 1]
+    count += +@verse if @verse?
```
Wasn't this already coerced to a Number earlier?

sortofbusy replied:
Yes, good call.

davewasmer commented on lib/reference.coffee:
```diff
@@ -80,7 +82,7 @@ class Reference
     for i in [1...@book]
       count += books[i - 1].verses.length
-    count += @chapter
+    count += +@chapter
```
Same here - isn't this already a number?

sortofbusy replied:
Changed.

davewasmer commented on lib/reference.coffee:
```diff
@@ -160,6 +172,7 @@ class Reference
     for i in [1...bookId]
       count += Reference.chaptersInBookId(i)
     return count
+    return 0
```
This line will never be reached, right?

sortofbusy replied:
Yes, I put that there while trying to figure out why I was getting a certain result. Taken out.

davewasmer commented on lib/reference.coffee:
```diff
@@ -153,6 +164,7 @@ class Reference
         count += verse
         chapter += 1
         return count if chapter is chapterId
+    return 0
```
When would this line be reached?

sortofbusy replied:
Handles the edge case of chapter 1. Otherwise, the function returns undefined.

davewasmer commented on lib/reference.coffee:
```diff
@@ -131,11 +133,20 @@ class Reference
+    chapter = 1
+    for book in books
+      for verse in book.verses
+        return verse if chapter is chapterId
```
This essentially counts up until you reach the chapter ID in question. A faster approach would be to count the number of chapters in a book (i.e. `book.length`) and subtract that from the id as you go, rather than iterating through each chapter of each book. Something like:

```coffee
for book in books
  if chapterId > books.length
    chapterId -= books.length
  else
    return books[chapterId]
# potentially throw an error or something here - this line outside the for loop 
# should be impossible to reach for valid references
```

sortofbusy replied:
Good suggestion: I used this method. Out of range returns undefined, which I'm ok with.

davewasmer commented:
@sortofbusy thanks so much for the PR! Good catch on those off-by-one errors, and thanks for the several other improvements as well.

I left a few comments above - mostly stylistic, nothing major. Feel free to ping me if you have any questions about them.

sortofbusy commented:
All done - I went through and incorporated your suggestions. I haven't done exhaustive testing but most cases should work fine. I also added some commits modifying the books and README files. Thanks for all the feedback! It's my first pull request so I'm still getting the hang of this.

sortofbusy commented:
@davewasmer not sure if you saw my latest comments.

davewasmer commented:
@sortofbusy thanks for updating the PR! I'll merge these changes in. I really need to find some time to write some tests for this next ...
